libopenage/unit/producer.cpp

```diff
@@ -25,27 +25,56 @@
 namespace openage {
 
 std::unordered_set<terrain_t> allowed_terrains(const gamedata::ground_type &restriction) {
-	std::unordered_set<terrain_t> result;
+	// returns a terrain whitelist for a given restriction
+    // you can also define a blacklist for brevity, it will be converted and returned
+    std::unordered_set<terrain_t> whitelist;
+    std::unordered_set<terrain_t> blacklist;
 
 	// 1, 14, and 15 are water, 2 is shore
 	if (restriction == gamedata::ground_type::WATER ||
 	    restriction == gamedata::ground_type::WATER_0x0D ||
 	    restriction == gamedata::ground_type::WATER_SHIP_0x03 ||
 	    restriction == gamedata::ground_type::WATER_SHIP_0x0F) {
-		result.insert(1);
-		result.insert(2);
-		result.insert(14);
-		result.insert(15);
-	}
-	else {
-		for (int i = 0; i < 32; ++i) {
-			if (i != 1 && i != 14 && i != 15) {
-				result.insert(i);
-			}
-		}
-	}
-
-	return result;
+        whitelist.insert(1); // water
+        whitelist.insert(2); // shore
+        whitelist.insert(4); // shallows
+        whitelist.insert(14); // medium water
+        whitelist.insert(15); // deep water
+    }
+    else if (restriction == gamedata::ground_type::SOLID) {
+        blacklist.insert(1); // water
+        blacklist.insert(4); // shallows
+        blacklist.insert(14); // medium water
+        blacklist.insert(15); // deep water
+    }
+	else if (restriction == gamedata::ground_type::FOUNDATION ||
+             restriction == gamedata::ground_type::NO_ICE_0x08 ||
+             restriction == gamedata::ground_type::FOREST) {
+        blacklist.insert(1); // water
+        blacklist.insert(4); // shallows
+        blacklist.insert(14); // medium water
+        blacklist.insert(15); // deep water
+        blacklist.insert(18); // ice
+    }
+    else {
+        log::log(MSG(warn) << "undefined terrain restriction, assuming solid");
+        blacklist.insert(1); // water
+        blacklist.insert(4); // shallows
+        blacklist.insert(14); // medium water
+        blacklist.insert(15); // deep water
+    }
+
+    // if we're using a blacklist, fill out a whitelist with everything not on it
+    if (blacklist.size() > 0) {
+        // Allow all terrains that are not on blacklist
+        for (int i = 0; i < 32; ++i) {
+            if (blacklist.count(i) == 0) {
+                whitelist.insert(i);
+            }
+        }
```

Comment:
Your indentation here and in the lines above is wrong, please have a look at [our guide](https://github.com/SFTtech/openage/blob/master/doc/code_style/tabs_n_spaces.md).